map_ttl.go

```diff
@@ -190,29 +190,46 @@ func (c *MapTTLCache[K, V]) Len() int {
 }
 
 func (c *MapTTLCache[K, V]) set(key K, value V) {
+	ts := c.now()
 	val := ttlRec[K, V]{
 		value:     value,
 		prev:      c.tail,
-		timestamp: c.now(),
+		timestamp: ts,
 	}
 
 	if c.head == c.zero {
 		c.head = key
 		c.tail = key
-		val.prev = c.zero
 		c.data[key] = val
 		return
 	}
 
+	// If it's already the tail, we only need to update the value and timestamp
```

Comment:
Trying to wrap my head around what actually was happening before. Looks like the root issue is that we were overwriting tail with record that had `prev = tail`?

Reply:
Updating head should  not be a common case when cache is heavily utilised. Updating tail will.

Reply:
But both cases were be buggy :)